portfolio/portfolio.go

```diff
@@ -233,84 +256,88 @@ func (b *Base) AddAddress(address, description string, coinType currency.Code, b
 // coinType
 func (b *Base) RemoveAddress(address, description string, coinType currency.Code) error {
 	if address == "" {
-		return errors.New("address is empty")
+		return errAddressIsEmptyOrInvalid
 	}
 
-	if coinType.String() == "" {
-		return errors.New("coin type is empty")
+	if coinType.IsEmpty() {
+		return currency.ErrCurrencyCodeEmpty
 	}
 
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
 	idx := slices.IndexFunc(b.Addresses, func(a Address) bool {
 		return a.Address == address && a.CoinType.Equal(coinType) && a.Description == description
 	})
 	if idx == -1 {
-		return errors.New("portfolio item does not exist")
+		return errPortfolioItemNotFound
 	}
 	b.Addresses = slices.Clip(slices.Delete(b.Addresses, idx, idx+1))
 	return nil
 }
 
 // UpdatePortfolio adds to the portfolio addresses by coin type
-func (b *Base) UpdatePortfolio(addresses []string, coinType currency.Code) error {
-	if strings.Contains(strings.Join(addresses, ","), ExchangeAddress) ||
-		strings.Contains(strings.Join(addresses, ","), PersonalAddress) {
+func (b *Base) UpdatePortfolio(ctx context.Context, addresses []string, coinType currency.Code) error {
+	if slices.ContainsFunc(addresses, func(a string) bool {
+		return a == PersonalAddress || a == ExchangeAddress
+	}) {
 		return nil
 	}
 
-	switch coinType {
-	case currency.ETH:
-		for x := range addresses {
-			result, err := b.GetEthereumBalance(addresses[x])
-			if err != nil {
-				return err
-			}
+	type balanceRetriever func(ctx context.Context, address string) (float64, error)
 
-			if result.Error.Message != "" {
-				return errors.New(result.Error.Message)
-			}
+	var providerName string
+	var getBalance balanceRetriever
```

Comment:
```suggestion
	var getBalance func(ctx context.Context, address string) (float64, error)
```

and remove the type dec `balanceRetriever` above